File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.swift
import Foundation

func solution(_ numbers:[Int], _ target:Int) -> Int {
    
    var answer = 0
    func dfs(_ num: Int, _ d: Int) {
        if d == numbers.count {
            if num == target { answer += 1 }
            return
        }
        
        dfs(num + numbers[d], d+1)
        dfs(num - numbers[d], d+1)
    }
    dfs(0, 0)
    
    return answer
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def dfs(numbers, target):
    answer = 0
    total = 0
    
    def select(k):
        nonlocal answer, numbers, target, total
        
        if len(numbers) == k:
            if total == target:
                answer += 1
            return
        
        total += numbers[k]
        select(k+1)
        total -= numbers[k]
        
        total -= numbers[k]
        select(k+1)
        total += numbers[k]
        
    select(0)
    
    return answer

def solution(numbers, k):
    return dfs(numbers, k)




# 숫자는 2~20개
# 2 ^ 20의 개수 -> 104만 > Safe